Fix dataflow warnings for invalid trim annotations

sbomer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/101211

sbomer committed:
Add check for invalid method parameter type

sbomer committed:
Same for fields

sbomer committed:
Same for return values

sbomer committed:
Fix check for annotations on substituted generic parameter

sbomer committed:
FIx behavior for substituted generic parameter

sbomer committed:
Fix for generic return value

sbomer committed:
Fix return value behavior for ILLink/ILC

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs:
```diff
@@ -191,7 +191,7 @@ void GetDiagnosticsForField (Location location, IFieldSymbol fieldSymbol)
 				ReportRequiresUnreferencedCodeDiagnostic (location, requiresUnreferencedCodeAttributeData, fieldSymbol);
 
-			if (fieldSymbol.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None) {
+			if (FlowAnnotations.GetFieldAnnotation (fieldSymbol) != DynamicallyAccessedMemberTypes.None) {
```
Should we change most of the uses of `ISymbol.GetDynamicallyAccessedMemberTypes()` to instead use `FlowAnnotations.GetXAnnotation()` (except for when we warn on annotations on unsupported types)? Or at least a doc comment on it pointing to the `FlowAnnotations` for most use cases.

Also, I think we might still have extra warnings about DAM annotation mismatch on base/override method generic parameters, and property definition and accessor.

https://github.com/dotnet/runtime/blob/7cd266e3ad282d94e19664c6078ee699b9a5cc3e/src/tools/illink/src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs#L223-L246

https://github.com/dotnet/runtime/blob/7cd266e3ad282d94e19664c6078ee699b9a5cc3e/src/tools/illink/src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs#L248-L254

sbomer committed:
PR feedback

- Add comment
- Fix remaining uses of GetDynamicallyAccessedMemberTypes
  outside of invalid annotation warnings

This adds some tests for more cases that are fixed now:
- Property definition and accessor mismatch with invalid annotations
- Virtual and override mismatch with invalid annotations
